aarch64/src/mailbox.rs

```diff
@@ -7,9 +12,80 @@ pub const MBOX_WRITE: u64 = 0x20;
 pub const MBOX_FULL: u32 = 0x8000_0000;
 pub const MBOX_EMPTY: u32 = 0x4000_0000;
 
+static MAILBOX: Lock<Option<&'static mut Mailbox>> = Lock::new("mailbox", None);
+
+/// Mailbox init.  Mainly initialises a lock to ensure only one mailbox request
+/// can be made at a time.  We have no heap at this point, so creating a mailbox
+/// that can be initialised based off the devicetree is rather convoluted.
+pub fn init(dt: &DeviceTree) {
+    static mut NODE: LockNode = LockNode::new();
+    let mut mailbox = MAILBOX.lock(unsafe { &NODE });
+    *mailbox = Some({
+        static mut MAYBE_MAILBOX: MaybeUninit<Mailbox> = MaybeUninit::uninit();
+        unsafe {
+            MAYBE_MAILBOX.write(Mailbox::new(dt));
+            MAYBE_MAILBOX.assume_init_mut()
+        }
+    });
+}
+
+/// https://developer.arm.com/documentation/ddi0306/b/CHDGHAIG
+/// https://github.com/raspberrypi/firmware/wiki/Mailbox-property-interface
+pub struct Mailbox {
+    reg: RegBlock,
+}
+
+impl Mailbox {
+    fn new(dt: &DeviceTree) -> Mailbox {
+        Mailbox {
+            reg: dt
+                .find_compatible("brcm,bcm2835-mbox")
+                .next()
+                .and_then(|uart| dt.property_translated_reg_iter(uart).next())
+                .and_then(|reg| reg.regblock())
+                .unwrap(),
+        }
+    }
+
+    fn request<T, U>(&self, req: &mut Message<T, U>)
+    where
+        T: Copy,
+        U: Copy,
+    {
+        // Read status register until full flag not set
+        while (read_reg(self.reg, MBOX_STATUS) & MBOX_FULL) != 0 {}
+
+        // Write the request address combined with the channel to the write register
+        let channel = ChannelId::ArmToVc;
+        let uart_mbox_u32 = ptr::addr_of!(req) as u32;
```

Comment:
Rust is smart, and we gotta be smarter :crab: :bulb: 

Sooo... deref the `req` here! I.e.,
```rs
let uart_mbox_u332 = ptr::addr_of!(*req) as u32;
```

before:
```
Exception return from AArch64 EL2 to AArch64 EL1 PC 0x80070

r9 from the Internet
DTB found at: 0xffff800008000000
Binary sections:
  boottext:     0xffff800000080000-0xffff800000085000 (0x5000)
  text:         0xffff800000200000-0xffff800000400000 (0x200000)
  rodata:       0xffff800000400000-0xffff800000600000 (0x200000)
  data:         0xffff800000600000-0xffff8000006040a0 (0x40a0)
  bss:          0xffff8000006040a0-0xffff800000800000 (0x1fbf60)
  total:        0xffff800000080000-0xffff800000800000 (0x780000)
Physical memory map:
  Memory:       0x0000000000000000-0x000000003f215040 (0x3f215040)
looping now
```

after:
```
Exception return from AArch64 EL2 to AArch64 EL1 PC 0x80070

r9 from the Internet
DTB found at: 0xffff800008000000
Binary sections:
  boottext:     0xffff800000080000-0xffff800000085000 (0x5000)
  text:         0xffff800000200000-0xffff800000400000 (0x200000)
  rodata:       0xffff800000400000-0xffff800000600000 (0x200000)
  data:         0xffff800000600000-0xffff8000006040a0 (0x40a0)
  bss:          0xffff8000006040a0-0xffff800000800000 (0x1fbf60)
  total:        0xffff800000080000-0xffff800000800000 (0x780000)
Physical memory map:
  Memory:       0x0000000000000000-0x000000003c000000 (0x3c000000)
looping now
```

Reply:
And even nicer:
```rs
let uart_mbox_u32 = req as *const _ as u32;
```

That gets us rid of the macro and the `use core::ptr;`.